File: src/main/java/com/example/the_machine/models/ModelRegistry.java
package com.example.the_machine.models;

import com.example.the_machine.exception.ModelException;
import com.example.the_machine.models.enums.AnthropicChatModelSpec;
import com.example.the_machine.models.enums.OpenAiChatModelSpec;
import com.example.the_machine.models.enums.OpenAiEmbeddingModelSpec;
import jakarta.annotation.PostConstruct;
import java.util.Arrays;
import java.util.Set;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import lombok.val;
import org.springframework.ai.anthropic.AnthropicChatModel;
import org.springframework.ai.chat.model.ChatModel;
import org.springframework.ai.openai.OpenAiChatModel;
import org.springframework.ai.openai.OpenAiEmbeddingModel;
import org.springframework.beans.factory.config.ConfigurableListableBeanFactory;
import org.springframework.context.ApplicationContext;
import org.springframework.context.ConfigurableApplicationContext;
import org.springframework.stereotype.Service;

@Slf4j
@Service
@RequiredArgsConstructor
public class ModelRegistry {

  private final ModelProperties modelProperties;
  private final ModelFactory modelFactory;
  private final ApplicationContext applicationContext;

  @PostConstruct
  private void init() {
    registerModelsFromConfiguration();
  }

  private void registerModelsFromConfiguration() {
    val beanFactory = ((ConfigurableApplicationContext) applicationContext).getBeanFactory();

    // Register OpenAI chat models
    if (modelProperties.getOpenAi() != null && modelProperties.getOpenAi().getChat() != null) {
      modelProperties.getOpenAi().getChat().keySet().forEach(modelSpec ->
          registerOpenAiChatModel(beanFactory, modelSpec));
    }

    // Register OpenAI embedding models
    if (modelProperties.getOpenAi() != null && modelProperties.getOpenAi().getEmbedding() != null) {
      modelProperties.getOpenAi().getEmbedding().keySet().forEach(modelSpec ->
          registerOpenAiEmbeddingModel(beanFactory, modelSpec));
    }

    // Register Anthropic chat models
    if (modelProperties.getAnthropic() != null
        && modelProperties.getAnthropic().getChat() != null) {
      modelProperties.getAnthropic().getChat().keySet().forEach(modelSpec ->
          registerAnthropicChatModel(beanFactory, modelSpec));
    }
  }

  private void registerOpenAiChatModel(ConfigurableListableBeanFactory beanFactory,
      OpenAiChatModelSpec modelSpec) {
    try {
      val model = modelFactory.createOpenAiChatModel(modelSpec);
      val modelName = modelSpec.getModelName();
      beanFactory.registerSingleton(modelName, model);
      log.info("Registered OpenAI chat model bean: {}", modelName);
    } catch (ModelException e) {
      log.error("Failed to register OpenAI chat model: {}", modelSpec, e);
    }
  }

  private void registerOpenAiEmbeddingModel(ConfigurableListableBeanFactory beanFactory,
      OpenAiEmbeddingModelSpec modelSpec) {
    try {
      val model = modelFactory.createOpenAiEmbeddingModel(modelSpec);
      val modelName = modelSpec.getModelName();
      beanFactory.registerSingleton(modelName, model);
      log.info("Registered OpenAI embedding model bean: {}", modelName);
    } catch (ModelException e) {
      log.error("Failed to register OpenAI embedding model: {}", modelSpec, e);
    }
  }

  private void registerAnthropicChatModel(ConfigurableListableBeanFactory beanFactory,
      AnthropicChatModelSpec modelSpec) {
    try {
      val model = modelFactory.createAnthropicChatModel(modelSpec);
      val modelName = modelSpec.getModelName();
      beanFactory.registerSingleton(modelName, model);
      log.info("Registered Anthropic chat model bean: {}", modelName);
    } catch (ModelException e) {
      log.error("Failed to register Anthropic chat model: {}", modelSpec, e);
    }
  }

  public ChatModel get(String modelName) {
    return applicationContext.getBean(modelName, ChatModel.class);
  }

  public Set<String> listAvailableModels() {
    val beanFactory = ((ConfigurableApplicationContext) applicationContext).getBeanFactory();
    return Arrays.stream(beanFactory.getBeanNamesForType(ChatModel.class))
        .collect(java.util.stream.Collectors.toSet());
  }

  // Type-safe enum-based access methods
  public OpenAiChatModel get(OpenAiChatModelSpec model) {
    return applicationContext.getBean(model.getModelName(), OpenAiChatModel.class);
  }

  public OpenAiEmbeddingModel get(OpenAiEmbeddingModelSpec model) {
    return applicationContext.getBean(model.getModelName(), OpenAiEmbeddingModel.class);
  }

  public AnthropicChatModel get(AnthropicChatModelSpec model) {
    return applicationContext.getBean(model.getModelName(), AnthropicChatModel.class);
  }
}

File: src/main/kotlin/com/example/the_machine/models/ModelRegistry.kt
package com.example.the_machine.models

import com.example.the_machine.exception.ModelException
import com.example.the_machine.models.enums.AnthropicChatModelSpec
import com.example.the_machine.models.enums.OpenAiChatModelSpec
import com.example.the_machine.models.enums.OpenAiEmbeddingModelSpec
import jakarta.annotation.PostConstruct
import mu.KotlinLogging
import org.springframework.ai.anthropic.AnthropicChatModel
import org.springframework.ai.chat.model.ChatModel
import org.springframework.ai.openai.OpenAiChatModel
import org.springframework.ai.openai.OpenAiEmbeddingModel
import org.springframework.beans.factory.config.ConfigurableListableBeanFactory
import org.springframework.context.ApplicationContext
import org.springframework.context.ConfigurableApplicationContext
import org.springframework.stereotype.Service

private val logger = KotlinLogging.logger {}

@Service
class ModelRegistry(
  private val modelProperties: ModelProperties,
  private val modelFactory: ModelFactory,
  private val applicationContext: ApplicationContext
) {

  @PostConstruct
  private fun init() {
    registerModelsFromConfiguration()
  }

  private fun registerModelsFromConfiguration() {
    val beanFactory = (applicationContext as ConfigurableApplicationContext).beanFactory

    // Register OpenAI chat models
    modelProperties.openAi?.chat?.keys?.forEach { modelSpec ->
      registerOpenAiChatModel(beanFactory, modelSpec)
    }

    // Register OpenAI embedding models
    modelProperties.openAi?.embedding?.keys?.forEach { modelSpec ->
      registerOpenAiEmbeddingModel(beanFactory, modelSpec)
    }

    // Register Anthropic chat models
    modelProperties.anthropic?.chat?.keys?.forEach { modelSpec ->
      registerAnthropicChatModel(beanFactory, modelSpec)
    }
  }

  private fun registerOpenAiChatModel(
    beanFactory: ConfigurableListableBeanFactory,
    modelSpec: OpenAiChatModelSpec
  ) {
    try {
      val model = modelFactory.createOpenAiChatModel(modelSpec)
      val modelName = modelSpec.modelName
      beanFactory.registerSingleton(modelName, model)
      logger.info("Registered OpenAI chat model bean: {}", modelName)
    } catch (e: ModelException) {
      logger.error("Failed to register OpenAI chat model: {}", modelSpec, e)
    }
  }

  private fun registerOpenAiEmbeddingModel(
    beanFactory: ConfigurableListableBeanFactory,
    modelSpec: OpenAiEmbeddingModelSpec
  ) {
    try {
      val model = modelFactory.createOpenAiEmbeddingModel(modelSpec)
      val modelName = modelSpec.modelName
      beanFactory.registerSingleton(modelName, model)
      logger.info("Registered OpenAI embedding model bean: {}", modelName)
    } catch (e: ModelException) {
      logger.error("Failed to register OpenAI embedding model: {}", modelSpec, e)
    }
  }

  private fun registerAnthropicChatModel(
    beanFactory: ConfigurableListableBeanFactory,
    modelSpec: AnthropicChatModelSpec
  ) {
    try {
      val model = modelFactory.createAnthropicChatModel(modelSpec)
      val modelName = modelSpec.modelName
      beanFactory.registerSingleton(modelName, model)
      logger.info("Registered Anthropic chat model bean: {}", modelName)
    } catch (e: ModelException) {
      logger.error("Failed to register Anthropic chat model: {}", modelSpec, e)
    }
  }

  fun get(modelName: String): ChatModel {
    return applicationContext.getBean(modelName, ChatModel::class.java)
  }

  fun listAvailableModels(): Set<String> {
    val beanFactory = (applicationContext as ConfigurableApplicationContext).beanFactory
    return beanFactory.getBeanNamesForType(ChatModel::class.java)
      .toSet()
  }

  // Type-safe enum-based access methods
  fun get(model: OpenAiChatModelSpec): OpenAiChatModel {
    return applicationContext.getBean(model.modelName, OpenAiChatModel::class.java)
  }

  fun get(model: OpenAiEmbeddingModelSpec): OpenAiEmbeddingModel {
    return applicationContext.getBean(model.modelName, OpenAiEmbeddingModel::class.java)
  }

  fun get(model: AnthropicChatModelSpec): AnthropicChatModel {
    return applicationContext.getBean(model.modelName, AnthropicChatModel::class.java)
  }
}